### PR TITLE
Explicit conversion of C strings to Fortran strings during gets

### PR DESCRIPTION
**Pull request by jayeshkrishna**
* Explicitly convert C strings to Fortran strings during get.
   This fix is required for gets to work correctly with the pgi
   compiler (and is compatible with other compilers)
* Refactor put and get routines to improve the error handling
   and provide better user warning/error messages
* More tests for put/get of text variables
* Enhancement to the testing framework utils to compare
   arrays of strings

### CODE REVIEW

**Comment by jayeshkrishna** [comment]
This PR should fix issues with gets with the pgi compiler on compy (https://github.com/E3SM-Project/E3SM/issues/3667)

File: src/flib/pio_support.F90 [hunk 1]
```diff
@@ -49,6 +51,134 @@ subroutine replace_c_null(istr, ilen)
...
+          call piodie(__PIO_FILE__, __LINE__, "Not enough space to copy back the dimension sizes of the text variable")
+        end if
+      end if
+
+      ! Get the size of the other dimensions
+      ! These dimensions determine the size of the string array
+      do i=2,ndims
+        ierr = PIOc_inq_dimlen(File%fh, dimids(ndims - i + 1), dim_sz)
+        if(ierr /= PIO_NOERR) then
+          call piodie(__PIO_FILE__, __LINE__, "Inquiring the lenghts of dimension", i, " for the text variable failed")
```
Comment (dqwu) [inline src/flib/pio_support.F90:166]: lenghts ==> lengths

File: src/flib/pionfget_mod.F90.in [hunk 2]
```diff
@@ -499,4 +490,204 @@ CONTAINS
...
+    do i=1,min(max_clen, max_flen)
+      fstr(i:i) = cstr(i)
+      if(fstr(i:i) == C_NULL_CHAR) then
+        fstr(i:i) = F_SPACE_CHAR
+        cstr_copied = .true.
+        exit
+      end if
+    end do
+    if((.not. cstr_copied) .and. (max_flen < max_clen)) then
+      ! We have exhaused the Fortran write buffer
```
Comment (dqwu) [inline src/flib/pionfget_mod.F90.in:547]: exhaused ==> exhausted (total 12 occurrences)

**Review by dqwu — COMMENTED** [review]
Looks good to me. After refactoring, replace_c_null subroutine is still called 5 times in the Fortran code. Should we also refactor this routine?
1) At least the 2nd optional param ilen can be removed (no longer being used).
2) Maybe we can inline it for the remaining 5 calls, and use similar code in Cstring2Fstring_{DIMS}d?

File: src/flib/pio_support.F90 (hunk 1 above)
Comment (jayeshkrishna) [reply src/flib/pio_support.F90:166]: ok, thanks, will change it

File: src/flib/pionfget_mod.F90.in (hunk 2 above)
Comment (jayeshkrishna) [reply src/flib/pionfget_mod.F90.in:547]: ok, thanks - will change it

**Comment by jayeshkrishna** [comment]
The old code in these files requires further refactoring (like you noted above), we might be able to get rid of the replace_c_null routine too. 

* There are some other put/get routines that need refactoring
* The Fortran interface code currently does not have the correct error handling mechanism.

I think we can do that as a separate issue/PR

**Comment by dqwu** [comment]
> The old code in these files requires further refactoring (like you noted above), we might be able to get rid of the replace_c_null routine too.
> 
>     * There are some other put/get routines that need refactoring
> 
>     * The Fortran interface code currently does not have the correct error handling mechanism.
> 
> 
> I think we can do that as a separate issue/PR
Sounds good, thanks.

**Review by dqwu — APPROVED** [review]

**Comment by jayeshkrishna** [comment]
As @dqwu noted the new tests were failing for the ADIOS iotype with this PR because some output files in the tests that were modified multiple times were always opened with the read-only option.
This issue has now been fixed in the latest commit

**Review by dqwu — APPROVED** [review]

**Pull request merged**